codeSnippets/snippets/auth-oauth-google/src/main/kotlin/com/example/Application.kt

```diff
@@ -0,0 +1,63 @@
+package com.example
+
+import io.ktor.application.*
+import io.ktor.auth.*
+import io.ktor.http.*
+import io.ktor.response.*
+import io.ktor.routing.*
+import io.ktor.client.*
+import io.ktor.client.engine.cio.*
+import io.ktor.client.features.json.*
+import io.ktor.client.request.*
+
+fun Application.main() {
+    val httpClient = HttpClient(CIO) {
+        install(JsonFeature)
+    }
+    install(Authentication) {
+        oauth("auth-oauth-google") {
+            urlProvider = { "http://localhost:8080/callback" }
+            providerLookup = {
+                OAuthServerSettings.OAuth2ServerSettings(
+                    name = "google",
+                    authorizeUrl = "https://accounts.google.com/o/oauth2/auth",
+                    accessTokenUrl = "https://accounts.google.com/o/oauth2/token",
+                    requestMethod = HttpMethod.Post,
+                    clientId = System.getenv("GOOGLE_CLIENT_ID"),
+                    clientSecret = System.getenv("GOOGLE_CLIENT_SECRET"),
+                    defaultScopes = listOf("https://www.googleapis.com/auth/userinfo.profile")
+                )
+            }
+            client = httpClient
+        }
+    }
+    routing {
+        var principal: OAuthAccessTokenResponse.OAuth2? = null
+        authenticate("auth-oauth-google") {
+            get("/callback") {
+                principal = call.authentication.principal()
+                call.respondRedirect("/hello")
+            }
+        }
+        get("/") {
+            call.respondRedirect("/callback")
```

Comment:
Why do you need this redirect?

Reply:
to quickly open the Google authorization page when a user runs an application and clicks `http://0.0.0.0:8080`

Reply:
It's kinda confusing for a reader that you use `/callback` route to authenticate and as a callback URL. I'd recommend having a separate `/login` route in the `authenticate` block and redirect a user there.